### DISCUSSION TITLE
Documentation on adding new output formats?

### DISCUSSION THREAD

**Discussion by evelant** [tweag/nickel#2107] [General]
yaml, toml, and json cover a lot of use cases but not all of them. How do you go about adding a custom export format? I may have missed it but I didn't see anything about it in the docs. There are many configuration formats out there beyond yaml/toml/json. Being able to output to esoteric or arbitrary formats would likely be very useful.

**Comment by jneem** [comment]
Exporting is implemented using serde; the export function is [here](https://github.com/tweag/nickel/blob/77f6e75376e4b8166284e13fe9d84d5c89697b9e/core/src/serialize.rs#L425). If you're talking about contributing extra formats to nickel itself, you'll want to extend the `ExportFormat` enum and add the serialization at the place I linked to above.

If you're talking about formats that are too niche to be worth supporting upstream, and your formats have serde support, then you'll want to imitate what the [export cli command](https://github.com/tweag/nickel/blob/77f6e75376e4b8166284e13fe9d84d5c89697b9e/cli/src/export.rs#L30) does: evaluate a program, and then serialize the resulting `RichTerm` using serde.

Another option, of course, is to use json as an intermediate representation.
